File: 08-HEAP-PRIORITY-QUEUE/Python/kth-largest-element-in-a-stream.py
from typing import List

# https://leetcode.com/problems/kth-largest-element-in-a-stream/

class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.heap = []

        for i in nums:
            self.add(i)

    def add(self, val: int) -> int:
        length = len(self.heap)
        if length < self.k:
            self.heap.append(val)
            i = length
            while i > 0:
                parent = (i - 1) // 2
                if self.heap[parent] <= self.heap[i]:
                    break
                self.heap[parent], self.heap[i] = self.heap[i], self.heap[parent]
                i = parent
        elif val > self.heap[0]:
            self.heap[0] = val
            i = 0
            n = self.k

            while True:
                smallest = i
                left = 2 * i + 1
                right = 2 * i + 2
                if left < n and self.heap[left] < self.heap[smallest]:
                    smallest = left

                if right < n and self.heap[right] < self.heap[smallest]:
                    smallest = right

                if smallest == i:
                    break

                self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
                i = smallest

        return self.heap[0]

File: 08-HEAP-PRIORITY-QUEUE/PHP/kth-largest-element-in-a-stream.php
<?php

// https://leetcode.com/problems/kth-largest-element-in-a-stream/

class KthLargest
{
    private array $heap = [];

    function __construct(private readonly int $k, array $nums)
    {
        foreach ($nums as $num) {
            $this->add($num);
        }
    }

    function add(int $val): int
    {
        $length = count($this->heap);
        if ($length < $this->k) {
            $this->heap[] = $val;
            $i = $length;
            while ($i > 0) {
                $parent = floor(($i - 1) / 2);
                if ($this->heap[$parent] <= $this->heap[$i]) {
                    break;
                }

                $tmp = $this->heap[$i];
                $this->heap[$i] = $this->heap[$parent];
                $this->heap[$parent] = $tmp;
                $i = $parent;
            }


        } else if ($val > $this->heap[0]) {
            $this->heap[0] = $val;
            $i = 0;
            $n = $this->k;

            while (true) {
                $smallest = $i;
                $left = 2 * $i + 1;
                $right = 2 * $i + 2;

                if ($left < $n && $this->heap[$left] < $this->heap[$smallest]) {
                    $smallest = $left;
                }

                if ($right < $n && $this->heap[$right] < $this->heap[$smallest]) {
                    $smallest = $right;
                }

                if ($smallest == $i) {
                    break;
                }

                $tmp = $this->heap[$i];
                $this->heap[$i] = $this->heap[$smallest];
                $this->heap[$smallest] = $tmp;

                $i = $smallest;
            }
        }

        return $this->heap[0];
    }
}